### DISCUSSION TITLE
Does zeCommandQueueSynchronize synchronize on completion?

### DISCUSSION THREAD

**Discussion by chsigg** [jjfumero/codeBlogArticles#2] [General]
Hi Juan,

Thanks for the articles and source on getting started with Level Zero.

I noticed that you time your GPU execution from `zeCommandListClose` to `zeCommandQueueSynchronize`.

https://github.com/jjfumero/codeBlogArticles/blob/559948088e1730d2b2ec10de20d7cb1df214c25b/june2021/levelZero_MxM/mxm.cpp#L200

I think a more accurate timing would be to measure from `zeCommandQueueExecuteCommandLists` with a fence to `zeFenceHostSynchronize` (see [example here](https://spec.oneapi.io/level-zero/1.2.13/core/PROG.html?highlight=zecommandqueuesynchronize#fences)).

I've only just started to look at Level Zero and I'm not even familiar with Vulkan, but IIUC `zeCommandQueueSynchronize` just guarantees that the command queue has 'shipped' and is ready for reuse. It does not guarantee that execution on the GPU has completed.

I noticed that host-synchronizing on an event without synchronizing the command queue first gives an error. So I'm a little surprised about the example linked above which does not seem to require queue synchronization for fences.

**Comment by jjfumero** [comment]
Hi @chsigg , 
   thank you for your interest and your suggestions.

That example was more like an intro to Level Zero. IMO, that way of measuring: 

```cpp
    std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
    VALIDATECALL(zeCommandListClose(cmdList));
    VALIDATECALL(zeCommandQueueExecuteCommandLists(cmdQueue, 1, &cmdList, nullptr));    
    VALIDATECALL(zeCommandQueueSynchronize(cmdQueue, std::numeric_limits<uint64_t>::max()));
    std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
```

it will include data transfers (if needed) + kernel time. 

Looking at the  spec for `zeCommandQueueSynchronize`:  https://spec.oneapi.io/level-zero/latest/core/api.html?highlight=zecommandqueuesynchronize#zecommandqueuesynchronize 

> timeout: [in] if non-zero, then indicates the maximum time (in nanoseconds) to yield before returning ZE_RESULT_SUCCESS or ZE_RESULT_NOT_READY; if zero, then immediately returns the status of the command queue; if UINT64_MAX, then function will not return until complete or device is lost. Due to external dependencies, timeout may be rounded to the closest value allowed by the accuracy of those dependencies.

I think the use of a fence also has the same effect here, it will measure data transfers (if any) + kernel time, plus I also agree it makes more sense. Perhaps this is a question for the level-zero maintainers. 

Both approaches, quite likely, will also contain the dispatch time from the driver of all of these calls. But, IMO, to compare against sequential code or any other parallel approach, this should be the way to go (unless comparing only the GPU kernel time). 

Regarding the GPU kernel time, I have another post in which I explain how to measure just the GPU kernel time using events:

https://github.com/jjfumero/codeBlogArticles/blob/559948088e1730d2b2ec10de20d7cb1df214c25b/september2021/timingGPUKernel/mxm.cpp#L262-L267 

The blog-entry is here: https://jjfumero.github.io/posts/2021/09/timers-with-level-zero/
